src/tools/language-server/language-service.ts

```diff
@@ -8,90 +8,105 @@
  * http://polymer.github.io/PATENTS.txt
  */
 
-import {ErrorCodes, Message, StreamMessageReader as VSCodeStreamMessageReader, StreamMessageWriter as VSCodeStreamMessageWriter} from 'vscode-jsonrpc';
-import {isNotificationMessage, isRequestMessage, isResponseMessage} from 'vscode-jsonrpc/lib/messages';
+import {ExecuteCommandParams, InitializeParams, IConnection, Diagnostic, DiagnosticSeverity, DidChangeTextDocumentParams, DidOpenTextDocumentParams, DidSaveTextDocumentParams, TextDocumentSyncKind, InitializeResult} from 'vscode-languageserver';
 
-import {handlers} from './handlers.js';
-import {jsonrpc, Logger, LanguageServiceOptions, LanguageServiceContext, camelCase} from './util.js';
+import {Logger, LanguageServiceOptions, LanguageServiceContext, camelCase, normalizeUri, uri2path} from './util.js';
+import {LspLoader} from './lspLoader.js';
+import {gatherDiagnostics} from './diagnostics.js';
 
 export class LanguageService {
-  reader: VSCodeStreamMessageReader;
-  writer: VSCodeStreamMessageWriter;
-  context: LanguageServiceContext;  // The state.
-  initialized = false;  // error / onclose should trigger shutdown message.
-  streaming = false;    // Client supports partialResult.
-
-  constructor(
-      reader: VSCodeStreamMessageReader,
-      writer: VSCodeStreamMessageWriter,
-      options: LanguageServiceOptions,
-      logger: Logger) {
-    this.reader = reader;
-    this.writer = writer;
-    this.context = {logger, options};
-
-    this.reader.listen(message => {
-      this.update(message).then((response: Message) => {
-        this.context.logger.info('Response: ', response ? response : 'No response.');
-        if (response) {
-          writer.write(response);
-        }
-      }).catch(e => this.context.logger.error(e));
-    });
+  connection: IConnection;
+  rootPath: string | undefined = undefined;
+  logger: Logger;
+  options: LanguageServiceOptions;
+  isInitialized = false;  // error / onclose should trigger shutdown message.
+
+  constructor( connection: IConnection, options: LanguageServiceOptions, logger: Logger) {
+    this.connection = connection;
+    this.options = options;
+    this.logger = logger;
+  }
+
+  start() {
+    // Register service manager handlers
+    this.connection.onInitialize(this.initialize.bind(this));
+    this.connection.onInitialized(this.initialized.bind(this));
```

Comment:
Who named this thing `onInitialized` concomitant with `onInitialize`, Code Complete has a specific thing on not naming methods/etc. with one-letter differences, the author even used to work at MSFT. `onPostInitialize` or somesuch... anyway, these things are easy to critique after the fact. 🐱

Reply:
Yes, this part of the spec is also little observed. Many servers will ignore the initialized call and just use the initialize (no d) call.